src/lua/serializer.c

```diff
@@ -803,6 +1097,7 @@ int
 tarantool_lua_serializer_init(struct lua_State *L)
 {
 	static const struct luaL_Reg serializermeta[] = {
+		{"__gc", luaL_destroy_serializer},
```

Comment:
It seems correct, but dangerous. I think we should explicitly write the rules for managing the serializer objects, because now there are not-very-obvious constraints.

The best C-Lua interoperability is reached for objects that have a reference counter. It can be allocated from C, pushed to Lua then (many times) and it can be destroyed as from C as well as from Lua (when the refcounter reached zero). The lifetime of the object in C and Lua is arbitrary in the case: just increment the refcounter, use and decrement it. Increment on push to Lua and decrement in the Lua's GC handler.

We're in another situation here. We can't create a serializer object from C and then push it to Lua as a userdata object. We can't create a serializer object in Lua and use it in C after its lifespan in Lua ends.

In fact, we have two separate kinds of serializer objects:

* Serializer objects created in C in an automatic or a dynamic storage and to be destroyed in C. The destroy function (`luaL_serializer_free_options`) must be called on them (including objects in the automatic storage).
* Serializer objects created in Lua in the storage provided by a userdata object. Destroyed automatically, when the userdata is collected. Can be used in C only during a call of a function that references this object (usually using a upvalue) or if the serializer is a kind of always-loaded module.

The convertion from a Lua-citizen to a C-citizen serializer (or vice versa) is possible, but it requires to use the explicit copy constructor: `luaL_serializer_copy_options`.

Reply:
At least, now we have some brief suggestions. Thanks!